### ISSUE TITLE
Reasoning for: prefer_for_elements_to_map_fromIterable

### ISSUE THREAD

**Issue by Merrit** [dart-lang/linter#3259]
[prefer_for_elements_to_map_fromIterable](https://github.com/dart-lang/linter/blob/658f497eef358e989e3c41d60bfcc30bfaea217b/lib/src/rules/prefer_for_elements_to_map_fromIterable.dart) says not to use Map.fromIterable:

> When building maps from iterables, it is preferable to use for elements.
> 
> **BAD:**
> ```dart
> Map<String, WidgetBuilder>.fromIterable(
>   kAllGalleryDemos,
>   key: (demo) => '${demo.routeName}',
>   value: (demo) => demo.buildRoute,
> );
> ```
> **GOOD:**
> ```dart
> return {
>   for (var demo in kAllGalleryDemos)
>     '${demo.routeName}': demo.buildRoute,
> };
> ```

No reasoning is provided.

It seems this lint would prefer me to do something like:

```dart
enum Trees {
  maple,
  oak,
  willow,
}

final Map<Trees, List<String>> fromForLoop = {
  for (var tree in Trees.values) tree: []
};
```

Instead of something like this:

```dart
final Map<Trees, List<String>> fromIterable = Map.fromIterable(
  Trees.values,
  key: (tree) => tree,
  value: (_) => [],
);
```

The `fromIterable` option seems to me easier to write and cleaner to read. So if there is some concrete reason for the lint, I think we should include that reasoning - such as performance concerns or what-have-you.

I also notice the provided examples seem to be Flutter, so perhaps the lint is intended only for widget-based code? I looked for an original PR regarding this lint to find the reasoning, but I couldn't find one.

**Comment by pq** [comment]
Yeah.  There wasn't a lot of external conversation about this one and kind of developed around the white-board.  I agree 100% that the docs and rationale could be improved.  

Also, to be perfectly honest this one was partly based on speculation about how folks might benefit from using ui-as-code and was done before the feature landed.  It's possible we got it wrong...  Certainly it's likely we've learned a bit more in the meantime that could at least better inform the docs.

Thanks for bringing this up.  I'll tag some folks who might have some reflections now that the feature has been out there for a while.  Feel free to chime in with any thoughts of your own!

/cc @goderbauer @munificent @a14n 

(Oh and for reference, here's the PR where these bits initially landed: https://github.com/dart-lang/linter/pull/1505 and the referenced spec has moved to [control-flow-collections feature spec](https://github.com/dart-lang/language/blob/master/accepted/2.3/control-flow-collections/feature-specification.md).)

**Comment by pq** [comment]
> Interested to hear others' thoughts on the matter and learn even more.

Me too!

/fyi @eernstg @lrhn who may have perspectives to share.

**Comment by lrhn** [comment]
The language does have this feature (`Map.fromIterable`), but please don't use it!

We would never have added `Map.fromIterable` today because we now have a *better* approach in `for`-elements.
We didn't back then.
Also, `Map.fromIterable` interacts badly with null safety. We really should just deprecate it.

There are probably exceptions, mainly where you do a larger computation in the callback functions which won't fit in a single expression, but they're going to be rare.
In 99.99% of cases, writing the map literal directly is just better *and* more readable:
```dart
final Map<Trees, List<String>> literalIterable = {
  for (var tree in Trees.values) 
      if (tree != Trees.maple) 
          tree: []
};
```

Which one is more readable is obviously going to be a contentious and subjective issue.

Which is more efficient is more clear-cut: The literal avoids allocating three closures (one for `where`, two for `fromIterable`), one extra `Iterable` (the `where` iterable wrapping the original list), and calling those two-to-three closures for each element. 
On top of that, an optimizing compiler would be able to recognize that you are iterating a constant list, and might iterate it more efficiently using `.iterator.moveNext()`.

The above can be compiled into something equivalent to a `for` loop with an `if` statement with a `map[tree] = []` assignment, completely straight-line code with no extra function calls and getting the full benefit of local compiler optimizations, where `Map.fromIterable` is most likely compiled once, completely generically, and does everything polymorphically.


And because `fromIterable` predates null safety, the `key`/`value` callbacks are optional parameters, even though they are only actually optional if the key/value type is non-nullable. It's a bad API in the current Dart.

**Comment by Merrit** [comment]
That's very interesting - thank you for the insight! :grinning: 

Given that there is an actual real-world reason to avoid `fromIterable`, you have my vote for depreciating it.

So might it be good to update this lint with the reasoning, something like "Poor performance, may be depreciated" and a few examples of good alternatives?

**Comment by munificent** [comment]
> ```dart
> final Map<Trees, List<String>> fromForLoop = {
>   for (var tree in Trees.values) tree: []
> };
> ```
>   
> Instead of something like this:
>
> ```dart
> final Map<Trees, List<String>> fromIterable = Map.fromIterable(
>   Trees.values,
>   key: (tree) => tree,
>   value: (_) => [],
> );
> ```

What is it that you prefer about the latter? The former is shorter and arguably simpler to reason about because it doesn't involve creating a couple of closures and then thinking about when they get invoked and with what arguments.

Note also that you could do:

```dart
final fromForLoop = {
  for (var tree in Trees.values) tree: <String>[]
};
```

Since this is a map literal with elements now, inference can more easily fill in the map type for you.

**Comment by Merrit** [comment]
> What is it that you prefer about the latter?

I think its the structure of it.

```dart
final Map<Trees, List<String>> fromForLoop = {
  for (var tree in Trees.values) tree: []
};
```

With this my brain always takes a second to say, ok what is happening? Oh, tree is the key and it maps to a list, okay.

***

```dart
final Map<Trees, List<String>> fromIterable = Map.fromIterable(
  Trees.values,
  key: (tree) => tree,
  value: (_) => [],
);
```

With this I think I like the named parameters - I often do prefer named. `key:` `value:` means my brain takes 0 time to zone in on the relevant bit.

***

It would be even better with the iterable being named:

```dart
final Map<Trees, List<String>> fromIterable = Map.fromIterable(
  iterable: Trees.values,
  key: (tree) => tree,
  value: (_) => [],
);
```

I also don't love that I can't control it visually, I do sometimes wish I could use braces or something if I wanted to make the code look neater / easier to read.

Incoming pseudocode:

```dart
final Map<Trees, List<String>> fromForLoop = {
  for (var tree in Trees.values) {
		key: tree, value: []
	}
};
```

This is all very nitpicky and probably just me though. :woman_shrugging: :smile: 

I am content to use the `for in` knowing there is a real reason to do so.
